### PR TITLE
Remove readWriteThis methods

### PR DESCRIPTION
**Pull request by benharsh**
This PR removes 'readWriteThis' methods from the compiler, modules, and tests. Relevant documentation has been updated in the following areas:
- ChapelIO module
- IO module
- "forwarding" technote
- "specialMethods" primer

The "NoRead" type in the modified example is changed from a class to a record so that the resulting error messages are cleaner.

Testing:
- [x] full local

### CODE REVIEW

**Comment by benharsh** [comment]
An example of the ``NoRead`` case before this change, if one were to modify it to call ``.read()``, which confusingly involes "AtomicT":
```
$CHPL_HOME/modules/standard/IO.chpl:3525: In function '_read_one_internal':
$CHPL_HOME/modules/standard/IO.chpl:3568: error: unresolved call 'AtomicT(int(64)).readThis(channel(false,dynamic,false))'
$CHPL_HOME/modules/standard/IO.chpl:793: note: this candidate did not match: qiovec_t.readThis(f)
$CHPL_HOME/modules/standard/IO.chpl:3568: note: because method call receiver with type 'AtomicT(int(64))'
$CHPL_HOME/modules/standard/IO.chpl:793: note: is passed to formal 'this: qiovec_t'
$CHPL_HOME/modules/standard/IO.chpl:3568: note: other candidates are:
$CHPL_HOME/modules/standard/IO.chpl:833: note:   iostyleInternal.readThis(f)
$CHPL_HOME/modules/standard/IO.chpl:951: note:   QioPluginFile.readThis(f)
note: and 69 other candidates, use --print-all-candidates to see them
  $CHPL_HOME/modules/standard/IO.chpl:3414: called as _read_one_internal(_channel_internal: qio_channel_ptr_t, param kind = 0: iokind, x: AtomicT(int(64)), loc: locale) from method '_readOne'
  $CHPL_HOME/modules/standard/IO.chpl:3627: called as (channel(false,dynamic,false))._readOne(param kind = 0: iokind, x: AtomicT(int(64)), loc: locale) from method 'readIt'
  $CHPL_HOME/modules/standard/ChapelIO.chpl:394: called as (channel(false,dynamic,false)).readIt(x: AtomicT(int(64))) from function 'readThisFieldsDefaultImpl'
  $CHPL_HOME/modules/standard/ChapelIO.chpl:380: called as readThisFieldsDefaultImpl(reader: channel(false,dynamic,false), type t = borrowed BaseLocale, x: borrowed BaseLocale, needsComma: bool) from function 'readThisFieldsDefaultImpl'
  $CHPL_HOME/modules/standard/ChapelIO.chpl:569: called as readThisFieldsDefaultImpl(reader: channel(false,dynamic,false), type t = borrowed AbstractLocaleModel, x: borrowed AbstractLocaleModel, needsComma: bool) from function 'readThisDefaultImpl'
  $CHPL_HOME/modules/internal/ChapelLocale.chpl:515: called as readThisDefaultImpl(reader: channel(false,dynamic,false), x: borrowed AbstractLocaleModel)
  within internal functions (use --print-callstack-on-error to see)
note: generic instantiations are underlined in the above callstack
```

With a record:
```
$CHPL_HOME/modules/standard/IO.chpl:3525: In function '_read_one_internal':
$CHPL_HOME/modules/standard/IO.chpl:3568: error: unresolved call 'NoRead.readThis(channel(false,dynamic,false))'
$CHPL_HOME/modules/standard/IO.chpl:793: note: this candidate did not match: qiovec_t.readThis(f)
$CHPL_HOME/modules/standard/IO.chpl:3568: note: because method call receiver with type 'NoRead'
$CHPL_HOME/modules/standard/IO.chpl:793: note: is passed to formal 'this: qiovec_t'
$CHPL_HOME/modules/standard/IO.chpl:3568: note: other candidates are:
$CHPL_HOME/modules/standard/IO.chpl:833: note:   iostyleInternal.readThis(f)
$CHPL_HOME/modules/standard/IO.chpl:951: note:   QioPluginFile.readThis(f)
note: and 69 other candidates, use --print-all-candidates to see them
  $CHPL_HOME/modules/standard/IO.chpl:3414: called as _read_one_internal(_channel_internal: qio_channel_ptr_t, param kind = 0: iokind, x: NoRead, loc: locale) from method '_readOne'
  $CHPL_HOME/modules/standard/IO.chpl:3876: called as (channel(false,dynamic,true))._readOne(param kind = 0: iokind, x: NoRead, loc: locale) from method '_readInner'
  $CHPL_HOME/modules/standard/IO.chpl:3896: called as (channel(false,dynamic,true))._readInner(args(0): NoRead) from method 'read'
  foo.chpl:16: called as (channel(false,dynamic,true)).read(args(0): NoRead)
```

**Review by mppf — APPROVED** [review]

File: modules/standard/ChapelIO.chpl [hunk 1]
```diff
@@ -38,74 +38,41 @@ of a Hello World program:
...
-In cases when the reading routine and the writing routine are more naturally
-separate, or in which only one should be defined, a Chapel program can define
-``readThis`` (taking in a single argument - a readable channel) and/or
-``writeThis`` (taking in a single argument - a writeable channel).
-
-If none of these routines are provided, a default version of ``readThis`` and
-``writeThis`` will be generated by the compiler. If ``readWriteThis`` is
-defined, the compiler will generate ``readThis`` or ``writeThis`` methods - if
-they do not already exist - which call ``readWriteThis``.
+A Chapel program can implement ''readThis'' and ''writeThis'' methods on a
```
Comment (lydia-duncan) [inline modules/standard/ChapelIO.chpl:46]: These render a little oddly for me, are they two single quotes in a row?  Should they be double quotes?

**Review by lydia-duncan — APPROVED** [review]
Awesome!  One minor nit, but I think this is a strict improvement overall

File: modules/standard/ChapelIO.chpl (hunk 1 above)
Comment (benharsh) [reply modules/standard/ChapelIO.chpl]: I pushed some changes that use double-backticks instead: ``readThis`` and ``writeThis``

**Pull request merged**